### ISSUE TITLE
`createStore` and `update` do not work with frozen objects

### ISSUE THREAD

**Issue by markerikson** [cafreeman/signalis#53]
Hi! I'm trying out Signalis after @chriskrycho  suggested I take a look at it.

For context, I maintain Redux, and also its associated Reselect memoized selector library.  

This last week I was playing around with using some "autotracking" approaches for memoization, based on `@glimmer/validator` and pzuraq's "minimal autotracking" examples:

- https://twitter.com/acemarke/status/1645270192698556419

Looking at Signalis, it seems very similar to the autotracking code I was borrowing.

I just tried out, but unfortunately I've run into an immediate blocker:

Signalis tries to add a `$STOREPROXY` symbol to target objects, and this throws an error if the object is frozen:

https://github.com/cafreeman/signalis/blob/fdbba0c5f5f0829815209ed14624e123b8275d20/packages/core/src/store.ts#L193

That's a problem in our case because Redux Toolkit uses Immer for immutable updates, and Immer auto-freezes all state recursively.

Here's a relatively minimal repro that shows this happening:

```ts
import { createStore, update } from '@signalis/core'

const obj1 = {
  value: 0,
}

const obj2 = {
  value: 1,
}

Object.freeze(obj1)
Object.freeze(obj2)

const signalisStore = createStore({
  state: obj1,
})

console.log('State 1: ', signalisStore.state.value)

update(signalisStore, (draft) => {
  draft.state = obj2
})

console.log(signalisStore.state.value)

/*
TypeError: Cannot define property Symbol(signal-store-proxy), object is not extensible
    at Function.defineProperty (<anonymous>)
    at X (D:\Projects\redux\temp\glimmer-tracked-test\node_modules\@signalis\core\dist\signalis-core.umd.cjs:1:7422)
    at Object.get (D:\Projects\redux\temp\glimmer-tracked-test\node_modules\@signalis\core\dist\signalis-core.umd.cjs:1:6766)
    at todo (d:\Projects\redux\temp\glimmer-tracked-test\main.ts:230:40)
*/
```

I'm pretty interested in seeing how Signalis performs compared to what I've hacked together (mostly by copy-pasting and fiddling with things). Any chance you could find a way to remove the symbol assignment requirement so it works with frozen values?

**Comment by cafreeman** [comment]
@markerikson Hi there! Thanks for opening up the issue, this is a good find. I think I've found a fairly straightforward fix.

> Any chance you could find a way to remove the symbol assignment requirement so it works with frozen values?

The symbol assignment is pretty deeply baked in to how stores work, so I'd like to avoid going down this path if possible. However, I think this should actually be fine.

The repro you gave (thanks btw) was blowing up specifically when you try to use `update` on a store that contains a frozen object (in other words, the actual store creation on the frozen object should be fine). When `update` does its thing, it unwraps the store, does its update stuff, and re-wraps it, which is where things were going haywire.

To work around this, I *think* we'll be okay to just have `unwrap` check to see if the thing it's unwrapping is frozen and then make a shallow copy. This should allow the rest of the store functionality to work as expected, but will have the one small tradeoff that any frozen object that gets unwrapped will no longer be referentially equal to the original object passed to `createStore`. That said, I don't imagine that will be an issue given that you're doing immutable stuff already, so referential equality wasn't a factor to begin with.

I'll cut a new release ASAP, and thanks again for the issue :)

**Comment by cafreeman** [comment]
Just published v0.0.12 containing these changes if you want to give them a shot.

> Does that mean that when we end up reading the final result of whatever derived / selected calculation in userland app code, that the returned values will no longer be the same references that were in the original data?

🤔 I believe it does mean that, since we'd need to shallow copy each frozen object/array during the unwrap. I'm curious to see if this ends up being an issue for your use case.

>  It works by creating a "storage / cell / tag" container instance for each level of wrapped object nesting, and storing the current value inside that container

This is (mostly) how Signalis stores work as well, but you're right that we do stash the actual proxy on the incoming object (and then remove it during `unwrap`). I can take a look at seeing what it would take to avoid it entirely though.

**Comment by markerikson** [comment]
Yeah, my gut says that making copies of all the original objects would likely be bad for our intended use case, although I can't prove it off the top of my head.

I'll give 0.12 a shot tonight to see if I can at least get the general selector implementation running.  If you can think of a way to sidestep the copying and keep the original data as-is, I'd definitely be interested in trying that out!

**Comment by cafreeman** [comment]
I'm super curious to hear more about your specific use case. If you find that this doesn't work and can come up with some kind of repro I'd love to take a look

**Comment by cafreeman** [comment]
Ah, this is really interesting. In all honesty, I don't think an actual Signalis store is the right fit here since it's really intended more as an abstraction over the other primitives that makes dealing with more complex data structures less cumbersome. 

I see what you're talking about with how `tracked-redux` very specifically creates a Redux-friendly store by completely divorcing the Proxy from the value itself, but that ends up being less of a general purpose (/ convenient) solution, which is what Signalis is aiming for here.

That said, I *do* think there's a way to do what you want using the other primitives, though it may look too similar to what you already have:

```ts
function sum(values: number[]): number {
  return values.reduce((acc, num) => {
    return acc + num;
  });
}

const store = {
  args: createSignal([]),
};

const cache = createDerived(() => {
  const result = sum.apply(null, store.args.value);
  return result;
});

function memoized() {
  if (store.args === arguments) {
    return cache.value;
  }

  // In real life, we should do this in a for loop over each element in arguments
  if (store.args.value[0] === arguments[0]) {
    console.log('arguments are the same, no recompute');
  } else {
    store.args.value = arguments;
  }

  return cache.value;
}

const input = Object.freeze([1, 2, 3]);

console.log(memoized(input));
console.log(memoized(input));
console.log(memoized(input));
console.log(memoized([...input, 4]));
console.log(memoized(input));
```

Which would print out:
```
6
arguments are the same, no recompute
6
arguments are the same, no recompute
6
10
6
```

However, looking at your original tweet thread, I saw this tweet https://twitter.com/acemarke/status/1645270206208507906 that mentioned a very specific use case (don't recompute a map over an array of Todo IDs when one of the Todos becomes complete). This *is* something you can absolutely do with Signalis, but I'm not yet sure if it's something that will work in your specific use case. Here's how I'd do that if I were just a dev building an app with Signalis:

```ts
const base = [
  {
    id: '1',
    complete: false,
  },
  {
    id: '2',
    complete: true,
  },
];

const store = createStore(base);

const ids = createDerived(() => {
  console.log('computing');
  return store.map((todo) => todo.id);
});

createEffect(() => {
  console.log(ids.value);
});

update(store, (draft) => {
  draft[0].complete = true;
});

update(store, (draft) => {
  draft.push({
    id: '3',
    complete: false,
  });
});

update(store, (draft) => {
  draft[2].complete = true;
});
```

Which will print out:
```
computing
[ '1', '2' ]
computing
[ '1', '2', '3' ]
```

In other words, it'll print once for the initial `createEffect` and then only print again when we actually add a new Todo, but changing the status of `complete` for any Todo doesn't trigger a recompute. It would also recompute if we happened to change the ID of an existing Todo.

I don't yet have a full solution for you, but I thought the stuff above might at least provide some useful insight. Maybe you'll see something in there you can run with.

I'll keep thinking about it and will let you know if I come up with anything else

**Comment by cafreeman** [comment]
This is great. I'm not at a computer right now, but looking at your code I'm guessing that the Signalis implementation just isn't memoizing anything (which seems to be confirmed by how slow it is). When you do `store.args = arguments` I suspect it's forcing that Derived to recompute every time since it's still just blowing away the old array and replacing it.

I'll see if I can come up with something and run it through your benchmark.

**Comment by markerikson** [comment]
Sure, thank you!

I just posted the current WIP over at https://github.com/reduxjs/reselect/pull/605 if you want to take a look.

**Comment by cafreeman** [comment]
@markerikson well, the bad news is i haven't *beaten* your benchmark, but I have gotten things a lot closer with some minor tweaks. Here's what I have so far:

```ts
import { createDerived, createSignal } from '@signalis/core'
import {
  createCacheKeyComparator,
  defaultEqualityCheck
} from './defaultMemoize'

const shallowEqual = createCacheKeyComparator(defaultEqualityCheck)

export function signalisMemoize<F extends (...args: any[]) => any>(func: F) {
  // we reference arguments instead of spreading them for performance reasons

  let lastArgs: IArguments | null = null
  const store = {
    args: createSignal<any[]>([])
  }

  const derived = createDerived(() => {
    const res = func.apply(null, store.args.value)
    return res
  })

  // console.log('Creating memoized function')
  function memoized() {
    if (!shallowEqual(lastArgs, arguments)) {
      let didUpdate = false
      const newArray = store.args.value.slice()

      for (let i = 0; i < arguments.length; i++) {
        if (store.args.value[i] !== arguments[i]) {
          didUpdate = true
          store.args.value[i] = arguments[i]
          newArray[i] = arguments[i]
        }
      }

      if (didUpdate) {
        store.args.value = newArray
      }

      lastArgs = arguments
    } else {
      // console.log('Same args: ', lastArgs, arguments)
    }
    // const start = performance.now()
    // console.log('Calling memoized: ', arguments)

    // const end = performance.now()
    // console.log('Memoized execution time: ', end - start)
    return derived.value
  }

  return memoized as F & { clearCache: () => void }
}
```

And here's the output:

```
Total recomputations:
cdCounters1 2001
cdCounters2 2001
cdTodoIds 1002
cdTodoIdsAndNames 1002
cdCompletedTodos 1002
cdCompletedTodos2 1002
caCounters1 2001
caCounters2 2001
caTodoIds 1002
caTodoIdsAndNames 1002
caCompletedTodos 1002
caCompletedTodos2 1002
Total elapsed times:  {
  defaultElapsed: 9.953250005841255,
  autotrackElapsed: 17.18275000154972
}
```

I'm sure there's a better way to do this that will bring that time down even more (namely making it so that updating slots in that array isn't completely terrible), but this was just with a little bit of poking around.

I'll keep poking at it this week :)

**Comment by markerikson** [comment]
Yeah, I think that actually is basically equivalent to how `defaultMemoize` works out of the box, which is a shallow equality check on the extracted input values (which show up here as `arguments` in `memoized`).  Definitely let me know what else you figure out!
